### ISSUE TITLE
Use enhanced enums

### ISSUE THREAD

**Issue by JosefWN** [shamblett/coap#80]
When Dart 2.17 is released, enhanced enums will simplify our code a bit: https://github.com/dart-lang/sdk/issues/47849

Examples:
- [x] Block option SZX could be an int enum
- [x] `CoapCode` could be broken down into int enums
- [x] `CoapMediaType` could be int enum
- [x] `CoapMessageType` could be int enum
- [ ] `Request.observe` could be int enum?
- [x] `CoapDefinedAddress` could be string enum?

**Comment by JKRhb** [comment]
Dart 2.17 has just been released by the way :) https://medium.com/dartlang/dart-2-17-b216bfc80c5d

**Comment by shamblett** [comment]
Yes, got it, FFI updates and experimental RISCV support are good news for me with my IOT hat on.

**Comment by JKRhb** [comment]
I just noticed that the option type numbers might also be an interesting place where the enhanced enums could be used. I'll try to come up with a PR for that soon.

**Comment by JKRhb** [comment]
I started working on refactoring the `CoapCode`s using enhanced enums by the way :)

**Comment by JosefWN** [comment]
A couple of minor things I noticed, which will likely resolve themselves with the re-write, but mentioning just in case. Not all codes are represented in `codeToString` (at least `continues` is missing), and the fallback doesn't show the code:

```dart
    if (isValid(code)) {
      if (isRequest(code)) {
        return 'Unknown Request [code {0}]'; // `$code` rather than `{0}`?
      // ...
```

**Comment by JKRhb** [comment]
> 
> - Block option SZX could be an int enum
> - CoapCode could be broken down into int enums
> - CoapMediaType could be int enum
> - CoapMessageType could be int enum
> - Request.observe could be int enum?
> - CoapDefinedAddress could be string enum?

I think from this list, we could mark `CoapCode`, `CoapMediaType`, and `CoapMessageType` as resolved now :) Should we maybe turn this issue into a task list for tracking the remaining candidates for enhanced enums?

**Comment by JosefWN** [comment]
Done! :)

**Comment by shamblett** [comment]
Package re released at version 5.0.0.

**Comment by shamblett** [comment]
Yes no.problem  a bit overly aggressive with my tidy up.

**Comment by JKRhb** [comment]
> Block option SZX could be an int enum

This is now also addressed by #122 and can be checked once the PR has been merged :)

**Comment by JosefWN** [comment]
LGTM!
